src/state_machine/transitions.cpp

```diff
@@ -0,0 +1,177 @@
+                            Sensors sensors_data, Motors motors_data)
+{
+  if (!checkEmergencyGeneral(log, embrakes_data, nav_data, batteries_data, telemetry_data,
+                             sensors_data, motors_data)) {
+    return NULL;
+  }
+  return FailureBraking::instance_;
+}
+
+//--------------------------------------------------------------------------------------
+// Module Status
+//--------------------------------------------------------------------------------------
+
+State *checkModulesInitialised(Logger &log, EmergencyBrakes embrakes_data, Navigation nav_data,
+                               Batteries batteries_data, Telemetry telemetry_data,
+                               Sensors sensors_data, Motors motors_data)
+{
+  if (!telemetry_data.calibrate_command) return NULL;
+
+  if (embrakes_data.module_status != ModuleStatus::kInit) return NULL;
+  if (nav_data.module_status != ModuleStatus::kInit) return NULL;
+  if (batteries_data.module_status != ModuleStatus::kInit) return NULL;
+  if (telemetry_data.module_status != ModuleStatus::kInit) return NULL;
+  if (sensors_data.module_status != ModuleStatus::kInit) return NULL;
+  if (motors_data.module_status != ModuleStatus::kInit) return NULL;
+
+  log.INFO("STM", "Calibrate command received and all modules initialised");
+  return Calibrating::instance_;
+}
+
+State *checkModulesReady(Logger &log, EmergencyBrakes embrakes_data, Navigation nav_data,
+                         Motors motors_data)
+{
+  // We're only checking Navigation, Motors and Embrakes because only those modules are doing
+  // calibration.
+  if (embrakes_data.module_status != ModuleStatus::kReady) return NULL;
+  if (nav_data.module_status != ModuleStatus::kReady) return NULL;
+  if (motors_data.module_status != ModuleStatus::kReady) return NULL;
+
+  log.INFO("STM", "All modules calibrated");
+  return Ready::instance_;
+}
+
+//--------------------------------------------------------------------------------------
+// Telemetry Commands
+//--------------------------------------------------------------------------------------
+
+State *checkLaunchCommand(Logger &log, Telemetry telemetry_data)
+{
+  if (!telemetry_data.calibrate_command) return NULL;
+
+  log.INFO("STM", "Launch command received");
+  return Accelerating::instance_;
+}
+
+State *checkShutdownCommand(Logger &log, Telemetry telemetry_data)
+{
+  if (!telemetry_data.shutdown_command) return NULL;
+
+  log.INFO("STM", "Shutdown command received");
+  return Off::instance_;
+}
+
+//--------------------------------------------------------------------------------------
+// Navigation Data Events
+//--------------------------------------------------------------------------------------
+
+State *checkEnteredBrakingZone(Logger &log, Navigation &nav_data, Telemetry &telemetry_data)
+{
+  // TODO(Franz): Fix braking buffer.
+  if (telemetry_data.run_length > nav_data.displacement + nav_data.braking_distance) return NULL;
```

Comment:
Also, I don't think it is related to this pull request (as this issue is from long time ago), but could you move `telemetry_data.run_length` to `nav_data`? I don't think it should belong to Telemetry.

Reply:
Yeah, I can do that.

Reply:
This is actually quite convenient for us, too. I thought about it yesterday but it wasn't relevant enough to investigate.